Sources/Sampler/metropolis_local_v2.hpp

```diff
@@ -0,0 +1,154 @@
+  inline void Read(Eigen::Ref<RowMatrix<double>> x) const noexcept;
+
+ private:
+  /// \brief Randomizes the state.
+  ///
+  /// New state is sampled uniformly from #local_states_
+  inline void RandomState();
+
+  /// \brief Randomizes the indices.
+  ///
+  /// New indices are sampled uniformly.
+  inline void RandomSites();
+
+  /// \brief Randomizes the values.
+  ///
+  /// New values are chosen uniformly from all possible quantum numbers except
+  /// the current one (to minimize staying in the same configuration).
+  inline void RandomValues();
+
+  /// \brief A vector of size `BatchSize()` which indices of sites at which we
+  /// suggest changing quantum numbers
+  Vector<Index> sites_;
+  /// \brief A vector of size `BatchSize()` which contains new (proposed)
+  /// quantum numbers.
+  Vector<double> values_;
+  /// \brief A matrix of size `BatchSize() x SystemSize()` with the
+  /// current state of #BatchSize() different Markov chains.
+  RowMatrix<double> state_;
+  /// \brief Allowed values for quantum numbers
+  std::vector<double> local_states_;
+
+  std::vector<Suggestion> proposed_;
+  DistributedRandomEngine engine_;
+};
+}  // namespace detail
+
+class MetropolisLocalV2 {
+  using InputType = RowMatrix<double>;
+  using ForwardFn = std::function<void(Eigen::Ref<const InputType>,
+                                       Eigen::Ref<Eigen::VectorXcd>)>;
+
+  RbmSpinV2& machine_;
+  detail::Flipper flipper_;
+  InputType proposed_X_;
+  Eigen::ArrayXcd proposed_Y_;
+  Eigen::ArrayXcd current_Y_;
+  Eigen::ArrayXd randoms_;
+  Eigen::Array<bool, Eigen::Dynamic, 1> accept_;
+
+  inline MetropolisLocalV2(RbmSpinV2& machine, Index batch_size,
+                           std::true_type);
+
+ public:
+  MetropolisLocalV2(RbmSpinV2& machine, Index batch_size);
+
+  Index BatchSize() const noexcept { return flipper_.BatchSize(); }
+  Index SystemSize() const noexcept { return flipper_.SystemSize(); }
+  RbmSpinV2& Machine() const noexcept { return machine_; }
+
+  std::pair<Eigen::Ref<const InputType>, Eigen::Ref<const Eigen::VectorXcd>>
+  Read();
+  void Next();
+  void Reset();
+};
+
+struct StepsRange {
+  StepsRange(std::tuple<Index, Index, Index> const& steps)
+      : start_{std::get<0>(steps)},
+        end_{std::get<1>(steps)},
+        step_{std::get<2>(steps)} {
+    CheckValid();
+  }
+
+  Index start() const noexcept { return start_; }
+  Index end() const noexcept { return end_; }
+  Index step() const noexcept { return step_; }
+  Index size() const noexcept { return (end_ - start_ - 1) / step_ + 1; }
+
+ private:
+  void CheckValid() const;
+
+  Index start_;
+  Index end_;
+  Index step_;
+};  // namespace netket
+
+std::tuple<RowMatrix<double>, Eigen::VectorXcd,
+           nonstd::optional<RowMatrix<Complex>>>
+ComputeSamples(MetropolisLocalV2& sampler, StepsRange const& steps,
+               bool compute_gradients);
+
+Eigen::VectorXcd LocalValuesV2(Eigen::Ref<const RowMatrix<double>> samples,
```

Comment:
Samplers should not hold localvalues/gradients, since the form of the gradient is dependent on the specific application (say variational versus supervised, unsupervised etc)